File: common/ekcommon/src/main/java/com/ekingdom/common/view/EKToolbar.java
package com.ekingdom.common.view;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.view.View;

import androidx.appcompat.widget.AppCompatImageView;
import androidx.appcompat.widget.AppCompatTextView;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.constraintlayout.widget.Group;

public class EKToolbar extends ConstraintLayout {
    AppCompatImageView imgBaseAction;
    AppCompatImageView imgBack;
    AppCompatImageView imgMore;
    AppCompatTextView tvTitle;
    AppCompatTextView tvMore;
    AppCompatTextView tvAction;
    Group groupBase;
    Group groupMore;
    private Drawable baseActionIconResource;
    private boolean isShowBaseActionIcon = false;
    private boolean isShowMoreIcon = false;
    private String title = "";
//    private OnToolbarActionClickListener onToolbarActionClickListener;

    public EKToolbar(Context context) {
        super(context);
    }
//
//    public EKToolbar(Context context, AttributeSet attrs) {
//        super(context, attrs);
//        init();
//        TypedArray typedArray = context.obtainStyledAttributes(attrs, R.styleable.EKToolbar, 0, 0);
//        try {
//            baseActionIconResource = typedArray.getDrawable(R.styleable.EKToolbar_baseActionItemIcon);
//            isShowBaseActionIcon = typedArray.getBoolean(R.styleable.EKToolbar_showBaseActionItem, isShowBaseActionIcon);
//            isShowMoreIcon = typedArray.getBoolean(R.styleable.EKToolbar_showMoreItem, isShowMoreIcon);
//            title = typedArray.getString(R.styleable.EKToolbar_title);
//        } catch (Exception e) {
//            Log.e(getClass().getSimpleName(), "EKToolbar: " + e.getMessage());
//        } finally {
//            typedArray.recycle();
//        }
//
//        if (baseActionIconResource != null) {
//            imgBaseAction.setImageDrawable(baseActionIconResource);
//        }
//        groupBase.setVisibility(isShowBaseActionIcon ? View.VISIBLE : View.GONE);
//        groupMore.setVisibility(isShowMoreIcon ? View.VISIBLE : View.GONE);
//        tvTitle.setText(title);
//    }
//
//    public void setOnToolbarActionClickListener(OnToolbarActionClickListener onToolbarActionClickListener) {
//        this.onToolbarActionClickListener = onToolbarActionClickListener;
//    }
//
//    private void init() {
//        setTag("Toolbar");
//        View.inflate(getContext(), R.layout.base_toolbar_item, this);
//        imgBaseAction = findViewById(R.id.imgAction);
//        imgMore = findViewById(R.id.imgMore);
//        imgBack = findViewById(R.id.imgBack);
//        tvTitle = findViewById(R.id.tvTitle);
//        tvMore = findViewById(R.id.tvMore);
//        tvAction = findViewById(R.id.tvAction);
//        groupBase = findViewById(R.id.groupBase);
//        groupMore = findViewById(R.id.groupMore);
//
//        tvTitle.setSelected(true);
//
//        findViewById(R.id.clBack).setOnClickListener((v) -> {
////            AndroidUtilities.preventTwoClick(v);
//            if (onToolbarActionClickListener != null)
//                onToolbarActionClickListener.onClickBack(v);
//        });
//        tvAction.setOnClickListener((v) -> {
////            AndroidUtilities.preventTwoClick(v);
//            if (onToolbarActionClickListener != null)
//                onToolbarActionClickListener.onClickBaseActionItem(v);
//        });
//        tvMore.setOnClickListener((v) -> {
////            AndroidUtilities.preventTwoClick(v);
//            if (onToolbarActionClickListener != null)
//                onToolbarActionClickListener.onClickMoreItem(v);
//        });
//    }
//
//    public void setShowBaseActionIcon(boolean isShow) {
//        this.isShowBaseActionIcon = isShow;
//        groupBase.setVisibility(this.isShowBaseActionIcon ? View.VISIBLE : View.GONE);
//    }
//
//    public void setBaseImageIcon(int drawableID) {
//        imgBaseAction.setImageResource(drawableID);
//    }
//
//    public void setTitle(String title) {
//        tvTitle.setText(title);
//    }
//
//    public interface OnToolbarActionClickListener {
//        void onClickBack(View view);
//
//        void onClickBaseActionItem(View view);
//
//        void onClickMoreItem(View view);
//    }
}

File: android/icafe_android/pcommon/src/main/java/com/ekingdom/common/view/EKToolbar.kt
package com.ekingdom.common.view

import android.content.Context
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import android.util.Log
import android.view.View
import android.view.ViewGroup
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.core.view.ViewCompat
import androidx.core.view.updateLayoutParams
import com.ekingdom.ekcommon.R
import kotlinx.android.synthetic.main.base_toolbar_item.view.*

class EKToolbar (context: Context?, attrs: AttributeSet?) : ConstraintLayout(context, attrs) {
    private var baseActionIconResource: Drawable? = null
    private var isShowBaseActionIcon = false
    private var isShowMoreIcon = false
    private var title = ""
    private var onToolbarActionClickListener: OnToolbarActionClickListener? = null

    init {
        init()
        val typedArray =
            context!!.obtainStyledAttributes(attrs, R.styleable.EKToolbar, 0, 0)
        try {
            baseActionIconResource =
                typedArray.getDrawable(R.styleable.EKToolbar_baseActionItemIcon)
            isShowBaseActionIcon = typedArray.getBoolean(
                R.styleable.EKToolbar_showBaseActionItem,
                isShowBaseActionIcon
            )
            isShowMoreIcon =
                typedArray.getBoolean(R.styleable.EKToolbar_showMoreItem, isShowMoreIcon)
            title = typedArray.getString(R.styleable.EKToolbar_title)!!
        } catch (e: Exception) {
            Log.e(javaClass.simpleName, "EKToolbar: " + e.message)
        } finally {
            typedArray.recycle()
        }

        if (baseActionIconResource != null) {
            imgAction.setImageDrawable(baseActionIconResource)
        }
        groupBase!!.visibility = if (isShowBaseActionIcon) View.VISIBLE else View.GONE
        groupMore!!.visibility = if (isShowMoreIcon) View.VISIBLE else View.GONE
        tvTitle!!.text = title
    }

    private fun init() {
        tag = "Toolbar"
        ViewCompat.setOnApplyWindowInsetsListener(this) { view, insets ->
            view.setPaddingTop(insets.systemWindowInsetTop)
            insets
        }
        View.inflate(context, R.layout.base_toolbar_item, this)
        clBack.setOnClickListener { v ->
            if (onToolbarActionClickListener != null) onToolbarActionClickListener?.onClickBack(v)
        }
        tvAction.setOnClickListener { v ->
            if (onToolbarActionClickListener != null) onToolbarActionClickListener?.onClickBaseActionItem(v)
        }
        tvMore.setOnClickListener { v ->
            if (onToolbarActionClickListener != null) onToolbarActionClickListener?.onClickMoreItem(v)
        }
    }

    fun setShowBaseActionIcon(isShow: Boolean) {
        isShowBaseActionIcon = isShow
        groupBase.visibility = if (isShowBaseActionIcon) View.VISIBLE else View.GONE
    }

    fun setBaseImageIcon(drawableID: Int) {
        imgAction.setImageResource(drawableID)
    }

    fun setTitle(title: String?) {
        tvTitle.text = title
    }

    fun setOnToolbarActionClickListener(onToolbarActionClickListener: OnToolbarActionClickListener?) {
        this.onToolbarActionClickListener = onToolbarActionClickListener
    }

    private fun View.setPaddingTop(value: Int) = updateLayoutParams<ViewGroup.MarginLayoutParams> {
        setPadding(paddingLeft, value, paddingRight, paddingBottom)
        height = pxFromDp(56f) + value
    }

    private fun pxFromDp(dp: Float): Int {
        return (dp * resources.displayMetrics.density).toInt()
    }

    fun getImageAction() = imgAction
    fun getImageBack() = imgBack

    interface OnToolbarActionClickListener {
        fun onClickBack(view: View?)
        fun onClickBaseActionItem(view: View?)
        fun onClickMoreItem(view: View?)
    }
}